api/object.rb

```diff
@@ -20,8 +20,13 @@ class Object < Google::YamlValidator
     # Represents an object that has a (mandatory) name
     class Named < Api::Object
       attr_reader :name
+      # original value of :name before the provider override happens
+      # same as :name if not overridden in provider
+      attr_reader :api_name
```

Comment:
@nelsonjr I know you also have the concept of `field` for properties. Field is optional and push that logic in the templates. I prefer this approach. Happy to rename or discuss this idea.